akka-stream/src/main/scala/akka/stream/impl/Throttle.scala

```diff
@@ -42,21 +41,27 @@ private[stream] class Throttle[T](cost: Int,
         else completeStage()
 
       override def onPush(): Unit = {
-        val timeElapsed = now() - previousTime
-        val currentTokens = Math.min(timeElapsed * speed + lastTokens, scaledMaximumBurst)
         val elem = grab(in)
         val elementCost = scale(costCalculation(elem))
-        if (currentTokens < elementCost)
-          mode match {
-            case Shaping ⇒
-              currentElement = Some(elem)
-              scheduleOnce(timerName, ((elementCost - currentTokens) / speed).millis)
-            case Enforcing ⇒ failStage(new RateExceededException("Maximum throttle throughput exceeded"))
-          }
-        else {
-          lastTokens = currentTokens - elementCost
-          previousTime = now()
+
+        if (lastTokens >= elementCost) {
+          lastTokens -= elementCost
           push(out, elem)
+        } else {
+          val currentTime = now()
+          val currentTokens = Math.min((currentTime - previousTime) * speed + lastTokens, scaledMaximumBurst)
+          if (currentTokens < elementCost)
+            mode match {
+              case Shaping ⇒
+                currentElement = Some(elem)
+                scheduleOnce(timerName, ((elementCost - currentTokens) / speed).nanos)
```

Comment:
> it will be zero instead of taking into account that the timer will almost always be late.

Ok, I didn't really care about that drift, but this can be solved by recording here the time and then instead of setting to zero, calculate the tokens that were accumulated since the delay.